client/battle/BattleRenderer.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * BattleFieldController.h, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#pragma once
+
+#include "../../lib/battle/BattleHex.h"
+
+class Canvas;
+class BattleInterface;
+
+enum class EBattleFieldLayer {
+					   // confirmed ordering requirements:
+	OBSTACLES     = 0,
+	CORPSES       = 0,
+	WALLS         = 1,
+	HEROES        = 1,
+	STACKS        = 1, // after corpses, obstacles
+	BATTLEMENTS   = 2, // after stacks
+	STACK_AMOUNTS = 2, // after stacks, obstacles, corpses
+	EFFECTS       = 3, // after obstacles, battlements
+};
+
+class BattleRenderer
+{
+public:
+	using RendererPtr = Canvas &;
+	using RenderFunctor = std::function<void(RendererPtr)>;

```

Comment:
std::function is the slowest way (and it can really impact performance). Don't know yet usage if RenderFunction, but if it's something in the loop, better to consider alternatives:
 - template 
 - function pointer

Reply:
I would rather use interface + virtual function here instead, but that would need some rework of classes that use this RenderFunction.
Overhead is obviously exists, but compared to execution time of functors it should be negligible